src/MonitorProfiler/Communication/IpcCommClient.cpp

```diff
@@ -0,0 +1,106 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "IpcCommClient.h"
+#include <memory>
+
+HRESULT IpcCommClient::Receive(IpcMessage& message)
+{
+    if (_shutdown.load())
+    {
+        return E_UNEXPECTED;
+    }
+    if (!_socket.Valid())
+    {
+        return E_UNEXPECTED;
+    }
+
+    //CONSIDER It is generally more performant to read and buffer larger chunks, in this case we are not expecting very frequent communication.
+    char buffer[sizeof(MessageType) + sizeof(int)];
+    int read = 0;
+    int offset = 0;
+
+    do
+    {
+        int read = recv(_socket, &buffer[offset], sizeof(buffer) - offset, 0);
+        if (read == 0)
+        {
+            return E_ABORT;
+        }
+        if (read < 0)
+        {
+#if TARGET_UNIX
+            if (errno == EINTR)
```

Comment:
Is there any risk of an infinite loop here? Do we need to have a counter of how many times we try again before we error out?

Reply:
I don't think it's a practical concern? The app would need to be continuously interrupted by signals.